packages/playwright-core/src/image_tools/compare.ts

```diff
@@ -0,0 +1,104 @@
+/**
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the 'License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { blendWithWhite, colorDeltaE94, rgb2gray } from './colorUtils';
+import { ImageChannel } from './imageChannel';
+import { ssim, FastStats } from './stats';
+
+const SSIM_WINDOW_RADIUS = 5;
+const VARIANCE_WINDOW_RADIUS = 1;
+
+function drawPixel(width: number, data: Buffer, x: number, y: number, r: number, g: number, b: number) {
+  const idx = (y * width + x) * 4;
+  data[idx + 0] = r;
+  data[idx + 1] = g;
+  data[idx + 2] = b;
+  data[idx + 3] = 255;
+}
+
+type CompareOptions = {
+  maxColorDeltaE94: number;
+};
+
+export function compare(actual: Buffer, expected: Buffer, diff: Buffer, width: number, height: number, options: CompareOptions) {
+  const {
+    maxColorDeltaE94
+  } = options;
+  const [r1, g1, b1] = ImageChannel.intoRGB(width, height, expected);
+  const [r2, g2, b2] = ImageChannel.intoRGB(width, height, actual);
+
+  const drawRedPixel = (x: number, y: number) => drawPixel(width, diff, x, y, 255, 0, 0);
+  const drawYellowPixel = (x: number, y: number) => drawPixel(width, diff, x, y, 255, 255, 0);
+  const drawGrayPixel = (x: number, y: number) => {
+    const gray = rgb2gray(r1.get(x, y), g1.get(x, y), b1.get(x, y));
+    const value = blendWithWhite(gray, 0.1);
+    drawPixel(width, diff, x, y, value, value, value);
+  };
+
+  const fastR = new FastStats(r1, r2);
```

Comment:
I'd make these 3 lazy, upon first non-matching pixel, because they probably take the most time of the happy path.